### ISSUE TITLE
Get size of header

### ISSUE THREAD

**Issue by theojepsen** [p4lang/p4-spec#660]
How do you get the size of a header in P4? Here is an example of what I'm trying to do:
```
action set_len() {
    hdr.ipv4.totalLen = len(hdr.ipv4) + len(hdr.udp) + len(hdr.my_proto);
}
```
`len` should return the size of the header (in bytes). In Python you can do this with Scapy. E.g. use `len(scapy.UDP())` to get the size of a UDP header (i.e. 8 bytes).

The best workaround that I've found is to use a `#define` to set all the header sizes at the begging of my program, e.g.
```
#define IP_HDR_SIZE  20
#define MY_PROTO_HDR_SIZE 12
```
However, this has caused my lots of headaches. I can't tell you how many times I've changed the number/size of header fields, forgetting to update the constant, causing me to waste hours troubleshooting parser errors.

This seems like something straight-forward to implement in the compiler, possibly in the front-end. Since this can be calculated statically at compile time, it should be target-agnostic.

**Comment by jnfoster** [comment]
What about variable length headers?

**Comment by theojepsen** [comment]
@jnfoster I was thinking about why this feature (get header size) is not part of the language, and variable length headers was the first thing that popped into mind. Obviously, this is not known statically at compile time.

Would it be possible to raise a compile-time exception if `len()` is called on variable length headers?

**Comment by mbudiu-vmw** [comment]
We could also define it to return the max size of the header. We probably should return the size in bits. An important question is whether we promise this to be a compile-time constant. This may cause some difficulties in the implementation since we need to know all types before we can compute widths, but we can probably work it out. I will try to see whether we can work it out.

Another question is whether this applies to types aka sizeof, or just to instances. In the latter case we can make it like a method call.

**Comment by jafingerhut** [comment]
Thoughts on this idea.  I am happy to create a PR along these lines, but wanted to write something up quickly that others may comment on:

+ The primary use case where I miss a feature like this is to get the length of a header in bytes, in order to know how much to add or subtract from a field like IPv4 Total Length, or IPv6 Payload Length, which are in units of bytes.  Therefore I think the minimum addition should be a way to get the length in bytes of a header with no varbit field.  Without that capability, I don't see much motivation to make a change.

+ Propose that whenever possible, these new operations return compile time constants, to enable compiler optimization of expressions involving multiple such values, e.g. adding up several header lengths, as shown in the example code of the first comment in this issue.

+ If a fixed length header is not a multiple of 8 bits long (which is allowed in P4_16 language spec, even though implementations may make that restriction, and several do), I would propose that the operation used to get the length of the header in bytes returns the minimum number of bytes required to hold that number of bits, e.g. if the length in bits is `B`, then the 'how many bytes long this header is' operation returns `floor((B+7)/8)`.

+ Since P4_16 allows implementations to have headers that are not multiples of 8 bits long, it seems like it may be reasonable to suggest that there be a separate operation that returns a header's length in bits.  This is a weakly offered suggestion more for completeness sake, rather than an operation that I know has some strong use case.  We could leave it out now and consider adding such a thing later, too.

+ Propose that this operation be valid on a single element of a header stack, too, and return the length in bytes of that one element of the header stack.  Also that it be useful on an _entire_ header stack, and return the total length of all elements of the header stack.

+ Propose that this operation be valid on a header_union, too, and in that case that it returns the maximum length among all elements of the header_union.

+ For headers with a varbit field in them, an operation that returns a compile time constant has only two reasonable options that I can think of: use 0 as the length of the varbit field, or use the varbit field's maximum length.  I would propose the maximum length as the more useful of the two.

+ Does anyone have interest as implementers in supporting an operation that returns the current run-time variable length of a varbit field?  That could be a separate issue from this one, if it runs into hurdles.  It is also something that is possible for a P4 programmer to use metadata to keep track of, since the varbit's length is known during parsing time.  I for one am happy to not include such a thing while resolving this issue.

+ Propose that this operation is _not supported_ on any other type of object, e.g. it is not supported on structs.  Why?  Because I think it will lead to interminable arguments about the meaning of such an operation, and I don't see a strong use case for it.

+ Syntax!  I dont' have strong opinions here, and am open to compiler writer's suggestions, e.g. especially if there is syntax that makes it easier for the compiler to treat the return value as a compile time constant.

Some possibilities for syntax:

+ sizeof(hdr_variable)
+ hdr_variable.size
+ hdr_variable.size()

If we support the ability to have a separate operation get the size in bits, I would propose including "bits" somewhere in its name, to distinguish it from the one that returns bytes.  We could also explicitly include "bytes" in the names of the operation that return bytes.

**Comment by mbudiu-vmw** [comment]
This looks very reasonable.
How about querying a type, and not a variable?

**Comment by mbudiu-vmw** [comment]
We should check that the grammar is not ambiguous then.

**Comment by jnfoster** [comment]
I'm not strongly opposed, but why query the type rather than the instance? We have no other methods on types. I guess I don't see a reason to invent that category of operation unless there is a deep reason to do so....

**Comment by jnfoster** [comment]
I suppose the familiarity with C's `sizeof` is another potential reason... although I personally find that borrowing features that work similarly to but not quite exactly the same as in another language is a recipe for confusion.

**Comment by mbudiu-vmw** [comment]
Actually using the syntax `.size` is a problem if the structure has a field named `size`...

**Comment by jafingerhut** [comment]
I wanted to add another comment to summarize some goals I have for discussing the addition of this feature, to hopefully streamline those discussions:

+ By restricting what this new feature can do to only return compile-time constants, it makes it straightforward to implement well before the back end of a compiler, and thus does not add any requirements in the data plane to _any_ P4 implementations.  It also enables compile-time simplification of expressions like `sizeof(hdr1) + sizeof(hdr2) + sizeof(hdr3)` down to a single constant like `38`.  I expect such expressions will be a fairly common use case for this feature, e.g. when adding multiple headers to a packet during tunnel encapsulation.

+ If there are any issues with extending the capability to type names, I would prefer to leave that out, at least for now.  Better to have a useful feature in the language, and limited to only working on variables with type header, than to further delay adding such a feature to the language.  It can be extended later if truly desired.

+ The previous comment applies to _any extensions over and above the functionality I propose above_.  I would prefer to move forward on a very delimited but useful feature, than stall on a more general thing.

Another proposal for syntax, if it avoids worries that `sizeof` will mislead people coming from C/C++:

+ `sizeBytes(hdr_variable)`

That makes the units immediately obvious to everyone, and a name like `sizeBits(hdr_variable)` another things that could be added with very clear units for the return value.

I suppose we could use `sizeOctets` instead of `sizeBytes` if people care much about the name octet vs. byte.  I think I'm too young to care about the difference, having never programmed a system where byte meant anything other than 8 bits :-)   https://www.differencebetween.com/difference-between-octet-and-vs-byte/

**Comment by mbudiu-vmw** [comment]
We should perhaps allow overloading: `hdr.sizeInBits` and `hdr.sizeInBits()`

**Comment by jafingerhut** [comment]
Following initial discussion during 2018-Dec-03 language design WG meeting on this issue, I volunteered to write a PR for the language spec with the details.  I will put a "stake in the sand" by using the syntax `sizeInByte(hdr_variable)`, but this could of course be changed.

**Comment by jafingerhut** [comment]
@liujed Jed, regarding your question during the meeting about not bothering to define this new operator on structs, it is really a combination of two reasons, not only the one I mentioned during the meeting:

+ I do not know of any _use case_ where someone would want to use it in a P4 program on structs.  The size of headers has a definite desired use case described in the original comment of this issue, and it is kind of strange that P4 doesn't already support it.

+ In addition to the limited (or perhaps 0) benefit, I believe there could be controversy in trying to find agreement on how to define it, and given the limited-or-0 benefit to coming to that agreement, I would rather avoid having those discussions.  I know the obvious most simple definition is: "sum up the size in bits of all struct elements, then convert that total to bytes", but what is the size of an error field in a struct?  An enum that does not have an underlying bit<W> type?  What if an implementation chooses to internally use padding bits when it stores the struct -- should those be taken into account?  Perhaps we could come to agreement on all of those issues, but for what possible gain, and how long would it take?  I would quickly lose patience in attaining the useful goal of defining a new operation on headers only.

**Comment by liujed** [comment]
I think it's fine to leave off structs for now. Later on, if there turns out to be a need to get the size of a struct, we can always add that. With #383 coming down the pipe, though, I can see it soon being useful to at least talk about the size of a bit-vector struct.

**Comment by jafingerhut** [comment]
But with the limit of sizeInByte() only implemented on header types, even if/when #383 is implemented, there would still only need to be a way to define the size of _header_ types that contain bit-vector structs.  So we still wouldn't need to get into discussions of issues like what the size of `error` and all `enum` types are, only the structs that can be inside of headers, which are intentionally limited.

**Comment by jafingerhut** [comment]
FYI, for those curious about the current p4c behavior if you try to use fields in a header with names of built-in methods like isValid, setValid, and setInvalid, it gives an error, at least if you try to use those fields as fields in a table key.  Attempting to compile the attached P4_16 program using the latest p4c as of 2018-Dec-07 gives these error messages:
```
% p4c --target bmv2 --arch v1model header-field-name-collision.p4
header-field-name-collision.p4(79): error: Key hdr.foohdr.isValid field type must be a scalar type; it cannot be bool function()
            hdr.foohdr.isValid : exact;
            ^^^^^^^^^^^^^^^^^^
header-field-name-collision.p4(80): error: Key hdr.foohdr.setValid field type must be a scalar type; it cannot be void function()
            hdr.foohdr.setValid : exact;
            ^^^^^^^^^^^^^^^^^^^
header-field-name-collision.p4(81): error: Key hdr.foohdr.setInvalid field type must be a scalar type; it cannot be void function()
            hdr.foohdr.setInvalid : exact;
            ^^^^^^^^^^^^^^^^^^^^^
```
I believe that this is reasonable behavior for p4c, by the way, and am not recommending that it be changed.  I am simply adding this comment to point out that if this is considered correct behavior, then adding new methods on headers will make those method names no longer usable as field names in headers.
[header-field-name-collision.p4.txt](https://github.com/p4lang/p4-spec/files/2659066/header-field-name-collision.p4.txt)

**Comment by mbudiu-vmw** [comment]
We could have the compiler allow overloading of these field names, under the assumption that the function field names must always be called. Since we don't have lambdas or function pointers this is reasonable for now.

**Comment by mbudiu-vmw** [comment]
Called = used within a MethodCallExpression.

**Comment by jafingerhut** [comment]
Created a draft PR here: https://github.com/p4lang/p4-spec/pull/712

**Comment by hesingh** [comment]
From the PR Andy created above in #712, there is one comment copied below.

"Maybe this should be done in conjunction with modules - we can then have stuff in core.p4 be in a module std, and you can perhaps say std.sizeInBits after you import std..."

Using std.sizeInBits is a perfect solution because it can be used anywhere in code such as in parser, deparser, control, and action.  However, the module system for p4c is one year away.

If the community cannot wait a year, here is a thought.  

Develop a new extern for use in core.p4.  Note core.p4 already uses the packet_in and packet_out externs.  I would define a new extern and use it for header size methods.  The name of the new extern is open to suggestions.  One obvious choice is "ext_header".  

For P4-16 code, the use would be like this:
bit<32> len = 0;
len = ext_header.sizeInBits(hdr);

Note, packet_in already uses methods such as advance and length.

**Comment by mbudiu-vmw** [comment]
The name of the extern will already polute the namespace.

**Comment by hesingh** [comment]
I see the requirement saying "use a variable with an example such as ```sizeInBits(h)``` where h is a header".   If the extern doesn't work, then, I don't see how we can support the example above.  

So, do we go back to ```type.sizeInBits()```?

**Comment by hesingh** [comment]
I am trying to understand the P4 program for ```type.size()```.  Would the program look like the code below?  The compiler can raise an error if size() is used with anything but a P4 header.

```
action set_len() {
    hdr.ipv4.totalLen = hdr.ipv4.size() + hdr.udp.size() + hdr.my_proto.size();
}
```

**Comment by mbudiu-vmw** [comment]
Yes, except that ipv4 is not a type, it is an expression.

**Comment by hesingh** [comment]
Ok.

However, internally in the compiler we use the expression to get to the ipv4 header and then use the size() operator.

**Comment by hesingh** [comment]
I have filed a new Issue to keep implementation details out of spec changes being discussed here.

[p4lang/p4c#1717](https://github.com/p4lang/p4c/issues/1717)

**Comment by hanw** [comment]
I came across a use case that might worth considering when designing the size operation on headers. The use case is kind-of target-specific. Nonetheless, I am hoping the proposal would not prohibit us from expressing the intended behavior in P4-16.

On certain target, it is possible that the compiler would try to re-adjust the layout of a header by adding padding bytes in the header (for whatever reason I am not going to explain). It is not the common case, but it does happen. 

```
header hdr {
    bit<7> field;
    bit<1> pad;
}
```
may become
```
header hdr {
    bit<7> field;
    bit<9> pad;
}
```

Now if we design P4-16 to have the sizeInBits() method on the type, it wouldn't work, because different instances of the same header may be treated differently by the compiler midend/backend. Each instance may have a different padding. Similarly, eliminating the sizeInBits() method in the frontend would not work, because the size is computed solely based on the syntax in the frontend code, thus prohibits the compiler to perform optimization on the header layout, which makes the sizeInBits() method unusable. (if used, the generated constant would be wrong).

I would propose the implement the sizeInBits() method on the header instance, just like what we did for isValid(), setValid() and setInvalid(). This way, the method is handled by individual target, and all we need to do in the frontend is to do the appropriate typechecking.

**Comment by mbudiu-vmw** [comment]
I am afraid that header padding will break many other passes.
If something like this is done, it should probably be done before the front-end.
We were talking about size being a compile-time constant, ideally substituted very early in the front-end, C-like.

**Comment by hanw** [comment]
Header padding is done in the backend. Frontend wouldn't know that padding ever happened, so it wouldn't break any frontend pass. From a user's perspective, the header layout is specified in the p4 program. It can be emitted, extracted according to the layout, and even advanced with the size of the header instance (symbolically). From the compiler's backend perspective, the layout that user provided for a header may not be the most optimal layout to program to the hardware, we need to allow compiler to optimize the layout. If the size of a header is substituted with a constant very early in the frontend, the compiler backend would have a very difficult time to undo the substitution and perform the optimization to header layouts.

For instance, if the sizeInBytes() method is substituted with a constant in the frontend.
```
header hdr {
   bit<7> field;
   bit<1> dummy;
}

pkt.advance(hdr.sizeInBytes());
```
would become
```
header hdr {
   bit<7> field;
   bit<1> dummy;
}

pkt.advance(1);
```

and if the compiler backend decides to optimize the layout to
```
header hdr {
   bit<7> field;
   bit<9> dummy;
}
/// the constant in advance() method would be wrong.
pkt.advance(1);
```

**Comment by hesingh** [comment]
I agree with Mihai.

Any backend padding/optimization that a target does, can be performed before the frontend as well, agree?

So, why do that and (a) save p4c from adding any padding, (b) support the sizeof feature which the community is asking for, and (c) p4runtime has an impact if padding is done on a header by the backend.

Thus, let's decide upon a schema/format for what a target wants to do towards padding etc. and do the work before the frontend.   thanks.

**Comment by hanw** [comment]
I don't agree with padding can be done before frontend. The information that the backend uses to decide what to pad and how many bits to pad is not available before the frontend. 

I am asking not to substitute sizeof method with a constant in the frontend. It is fine by me to do the substitution in a midend pass.

**Comment by hesingh** [comment]
What do you do with P4Runtime which flattens headers before midend is run?  

I would like to understand better why padding needs to happen after frontend.  I see all the frontend passes in ```p4c/frontends/p4/frontend.cpp```.  Off the cuff, I don't see a pass that has impact on header or struct in terms of changing sizes.  It would help to articulate what frontend passes operate that help a backend figure out how many bits to pad.

**Comment by hanw** [comment]
In the use case that I cam across, the header that is padded does not go to the control plane. So flattening in P4Runtime has no impact. 

With respect to why we need to add padding, I am afraid that involves hardware specifics, which I cannot say too much about in an open-forum like p4c repo. 

None of the frontend pass is going to help the backend to figure out how many bits to pad. It is purely a target-specific backend thing, and requires some target-specific analysis on hardware resource, etc. All I am asking is do not transform the program in a way that make certain backend difficult to write, and leave the choice to the backend implementer.

**Comment by jafingerhut** [comment]
Out of curiosity, is this header going out of the switch on a link, to another device?  If so, that seems like kind of unpredictable behavior that the other device may or may not know how to parse, yes?

I am guessing you have already considered the consequences, and have reasons to go forth with this plan, so will not argue too hard against you, but I just wanted to point out that the P4_16 spec has had the following sentence in section "Header types" since before Oct. 2016, when it was added in the git commit log: "There is no padding or alignment of the header fields."

**Comment by hanw** [comment]
@jafingerhut This header does not go out of the switch. It's the user-defined header that goes from ingress deparser to egress parser.

**Comment by mbudiu-vmw** [comment]
How is the user supposed to use this function in your setting? If we called this function declaredSizeInBits would you be happy? If in your architecture headers change sizes and the user needs to know when and where you should just define an architecture-specific function for the user to fond out. That should not preclude us from defining this function in the language definition.

**Comment by hesingh** [comment]
> I think it's fine to leave off structs for now. Later on, if there turns out to be a need to get the size of a struct, we can always add that. With #383 coming down the pipe, though, I can see it soon being useful to at least talk about the size of a bit-vector struct.

From day one, p4c/ir/type.def supports an implementation for width_bits() which supports nested structs for width computation.   The new code for sizeof in p4c sent out for PR uses width_bits() to compute the sizeof.

**Comment by hanw** [comment]
After pondering on this a bit more, I think what I need is the `serializable struct` concept that AndyKeep(?) proposed before. What I need is something that's like a header, because it is serialized/de-serialized, but it is not a header, because the layout might be changed by the compiler. It is also not a struct, because it can only have serializable types in it.

In the P4-16 spec, we could say that the size of a serializable struct cannot be determined by the frontend. It must be decided by the compiler midend or backend. The size of a 'pure' header is decidable in the frontend, except the header that has a serializable struct inside and the header that has a varbit in it.

Now if we allow header to have a nested serializable struct, we probably do not need header to have a nested struct anymore. The nested struct support in the frontend is basically implementing serializable struct, as it checks for `Type_Struct` and checks the members in the struct to be types that can be serialized. This can potentially simplify how to write the spec for nested struct in P4-16. There will be less confusion on what can be nested with what. 

At this moment, if we allow header to nest struct, then we can have 
- struct inside a struct,
- struct inside a header and the struct can only have bit types in it,
- header inside a struct,
- **but not** header inside a header

It raises more questions like: 
Can we have a header that nests a struct that nests a header?

So it is basically a mess.

If we use serializable struct, then the definition of nesting would be cleaner.
- struct inside a struct.
- header inside a struct.
- serializable struct inside a header.

We had a proposal for serializable struct, maybe it is the time to revive that discussion in LDWG?

**Comment by hesingh** [comment]
Mihai is gone for two weeks. He should weigh in first.  Recall, in the nested structs discussion, it has been said creating a new IR for every new type of feature such as nested struct does not make sense.  A new serializable struct would need new IR. 
 
> At this moment, if we allow header to nest struct, then we can have
> 
> * struct inside a struct,
> * struct inside a header and the struct can only have bit types in it,
> * header inside a struct,
> * **but not** header inside a header
> 
> It raises more questions like:
> Can we have a header that nests a struct that nests a header?
> 
No.  The onlyBitsOrBitStructs() check will fail.  The check is very clear to only allow serializable fields. 

> So it is basically a mess.

> We had a proposal for serializable struct, maybe it is the time to revive that discussion in LDWG?

IMHO, the Feb. 25th LDWG should first discuss any pending issues with recirculate etc. and this sizeof issue.

**Comment by jafingerhut** [comment]
Han, just a quick question -- I do not know all of the constraints you may be trying to satisfy all at once, but here is a copy of most of your first paragraph: "What I need is something that's like a header, because it is serialized/de-serialized, but it is not a header, because the layout might be changed by the compiler. It is also not a struct, because it can only have serializable types in it."

Why not use a struct, and if the struct is used in this particular way you have in mind, and it has field in it with types you don't support for that use case, you give an error message for that use?

Alternately, for your use case it sounds like you _could_ allow types like error and enum (without a base/backing type), because the struct is being passed from inside the device to a destination inside the device, so you can use whatever custom representation representation you want for those types and it will work for your target?

**Comment by hanw** [comment]
@jafingerhut Struct would not work, because a struct like below cannot be emitted or extracted.
```
struct mystruct_t {
   bit<9> field;
}

parser (mystruct_t st) {
   state start {
       pkt.extract(st);
   }
}
```

This on the other hand, could be legal.
```
serializable struct mystruct_t {
   bit<9> field;
}

parser (mystruct_t st) {
   state start {
       pkt.extract(st);
   }
}
```

> No. The onlyBitsOrBitStructs() check will fail. The check is very clear to only allow serializable fields.

@hesingh Precisely! The onlyBitsOrBitStructs is basically checking if a struct is a serializable struct. Why not just call the IR node IR::Type_SerStruct? Then the validator functor on line 1356 in typeChecker.cpp becomes
```
auto validator = [this] (const IR::Type* t) {
    while (t->is<IR::Type_Newtype>())
        t = getTypeType(t->to<IR::Type_Newtype>()->type);
        return t->is<IR::Type_Bits>() ||  t->is<IR::Type_Varbits>() || 
                   t->is<IR::Type_SerStruct>() || t->is<IR::Type_SerEnum>() || 
                   t->is<IR::Type_Boolean>(); }
```

I don't see any problem with introducing a new IR node if that makes the meaning of the IR more precise.

**Comment by jafingerhut** [comment]
So echoing something that Mihai suggested earlier -- suppose you use a header that can silently have padding added in the target-specific back end for your use case.  Does it _matter_ if a sizeBits() operator returned the unpadded length for your use case?  Who would _use_ the operator for such an internal-to-the-device-only header?  Certainly they wouldn't use it for sending out on a wire, and the main use case I know of for the sizeBits() / sizeBytes() operators are for people doing tunnel encapsulation and decapsulation for packets sent on ports to other devices.

**Comment by hanw** [comment]
Knowing the size of the "internal header" is useful if the program needs to recompute the size of the packet in egress pipeline. Typically, egress pipeline has access to some metadata that represents the length of the packet (could be original packet size at beginning of ingress, could be the modified packet size after ingress processing, it really depends on the data path that the packet goes through). If the program needs to update the ipv4 length field in egress, it needs to know about the size of the 'internal header', because the intrinsic metadata that count the packet size may/may not accounts for the 'internal header', we need the flexibility to express this packet length computation in P4.

**Comment by jafingerhut** [comment]
OK, another thought, and hopefully I can stop thinking about this for at least the weekend :-)

You want something like a header, in that it can have fields inside of it of various types, and you want to be able to emit() it and get out a target-specific sequence of bits, and later extract() the result on that target-specific sequence of bits and get back the original contents.

Sounds to me like a struct fits the bill.   Probably not as implemented in the current p4c and bmv2, but the idea has these advantages:

+ The P4_16 spec says very little today about the behavior of emit and extract on structs -- the only thing it says so far is that if a struct contains only headers, header stacks, and header_union elements, then emit must work by iterating emit() over those fields, in order.  We could explicitly say in the spec that emit() on structs that contain other types of fields can emit any implementation-specific sequence of bits that the target implementers choose, as long as a later extract will get back the original values of all fields.

+ Since it is target-specific, so is the size of the sequence of bits emitted.  So is the order of bits.  So is whether they are gzip'ed or not (joking, but you get the idea).

Unless I complete misinterpreted what the serializable struct goals were, I thought that one of them was trying to make the sequence of bits that were produced from emit to be _defined by the spec, portably across targets_.  That is 100% not the goal with what you are doing.

**Comment by hanw** [comment]
[serializable.pdf](https://github.com/p4lang/p4-spec/files/2847355/serializable.pdf)
@jafingerhut  I looked up the proposal for serializable type (attached). There are a couple paragraphs on page 2 that is relevant to this discussion.

> Where serializable data is used elsewhere, such as tuples passed in to an extern like the IP checksum extern, we expect that data to be packed-bit data, without any padding. In particular, tuples and structs with multiple fields are expected to be read and written without any padding bits. In the case of headers, if they are treated as serializable types they will need to read and write the validity information along with the header data.
> ...
> It is worth noting that none of these recommendations are intended to force implementers to use a particular representation for data. A CPU target, for instance, might want to specify aggregate data types with padding to better match the memory model on the CPU. Or an implementation might decide to completely flatten and re-arrange data to better fit a bus between parts of a hard- ware device. The intention is to provide some flexibility to the P4 programmer.
> ...
> Serializable types need not be represented as in their serializable format internal to the data plane. The serializable layout is only required for reading from the wire or writing to the wire, or in other extern functions or methods that specify a serializable modifier.
> 7.1. Why might the compiler decide on a different layout?
In some cases the serialized layout might impact performance when referenc- ing or updating data. For instance, on a CPU, extracting bit fields, may require shifting and masking to extract a field. In cases where the field is accessed frequently, it might make more sense to choose an internal representation that is more efficient for referencing the field, and then do the work necessary to serialize this when the serialized layout is needed.
It is up to the compiler to determine what the representation should be, the serializable indicator only requires that when the value is passed to an extern function or method requiring the serialized data that it be serialized.

In fact, it seems the proposal on serializable struct fits quite nicely to my use case.

**Comment by hesingh** [comment]
> @jafingerhut Struct would not work, because a struct like below cannot be emitted or extracted.
> 
> ```
> struct mystruct_t {
>    bit<9> field;
> }
> 
> parser (mystruct_t st) {
>    state start {
>        pkt.extract(st);
>    }
> }
> ```
> This on the other hand, could be legal.
> 
> ```
> serializable struct mystruct_t {
>    bit<9> field;
> }
> 
> parser (mystruct_t st) {
>    state start {
>        pkt.extract(st);
>    }
> }
> ```

Even when a struct is a new serializable struct, how does extract work when the serializable struct is inside a header or another serializable struct?  The serializable struct is extracted, but other bits in the outer header/struct are not.  Is the isValid bit set on the outer header/struct?  You have a comment in https://github.com/p4lang/p4-spec/issues/383 that says "changing extract to support a struct is a fundamental change to P4-16."  Or are you saying a serializable struct will never be nested?  If so, you have added another wrinkle.

You use an internal header for internal processing; the header is not emitted.  So why do you care if the internal header has emittable fields?   **Its your internal header, do what you want with it, including using an internal sizeof with this header.**  It's interesting to propose changes to a language for an asic internal representation of data store.  

Currently, p4c does not support sizeof.  I think the Tofino supports tunnel decap+encap, doesn't it?  Without sizeof, how is that working?  I don't see why the same internals won't work for ingress deparser to egress parser path.

**Comment by jafingerhut** [comment]
Without sizeof, tunnel encap and decap code sometimes requires P4 program authors to #define constants equal to the lengths of headers they have defined, and the values of those constants must be calculated by hand, and if the header definition ever changes, the value of the constant does not automatically change as a result.

**Comment by jafingerhut** [comment]
A struct like you describe _could_ be emitted and extracted, if we define a way for it to happen -- a way that has not been specified yet.  One way to do it would be to explicitly say "it is target-dependent how it is emitted and extracted, and extracting it is only guaranteed to work _on the same device_ where it was emitted, with no portable guarantees whatsoever, if you attempt to emit such a struct on a physical port leading to another device, so we strongly recommend against trying to do that.  Here is the gun.  Try not to hit your foot too many times."

**Comment by hesingh** [comment]
> Without sizeof, tunnel encap and decap code sometimes requires P4 program authors to #define constants equal to the lengths of headers they have defined, and the values of those constants must be calculated by hand, and if the header definition ever changes, the value of the constant does not automatically change as a result.
> 
> There is no road block to writing tunnel encap/decap code without a sizeof operator. There is an annoying inconvenience that I hope we can eliminate without writing a Ph.D. thesis on how to do it.

The annoying inconvenience has been dealt with in two days of coding with p4c to support sizeof.   My point was, I haven't seen a need yet to change the language to define a new serialiazable struct.

**Comment by hesingh** [comment]
@hanw The internal data store can be a header.  Why do you say, you cannot use a header because the header cannot be modified.  At packet ingress into the asic, copy the packet header and use the copy in the pipeline.  At URW stage, URW modify/insert tables help reconcile the copy and the original header.

**Comment by hanw** [comment]
@hesingh 
> Even when a struct is a new serializable struct, how does extract work when the serializable struct is inside a header or another serializable struct? The serializable struct is extracted, but other bits in the outer header/struct are not. Is the isValid bit set on the outer header/struct? 
This case is not unique to serializable struct, it happens with struct in a header as well.

> You have a comment in #383 that says "changing extract to support a struct is a fundamental change to P4-16." Or are you saying a serializable struct will never be nested? If so, you have added another wrinkle.
If you read the proposal for serializable type in the attached pdf [serializable.pdf](https://github.com/p4lang/p4-spec/files/2848368/serializable.pdf). The question is stated on page 6. We need to specify what nested serializable struct means, I imagine it would be similar to nested struct.

> You use an internal header for internal processing; the header is not emitted. So why do you care if the internal header has emittable fields? **Its your internal header, do what you want with it, including using an internal sizeof with this header.** It's interesting to propose changes to a language for an asic internal representation of data store.
The header is emitted and extracted. In this aspect, the serializable struct is similar to a header. However, the internal representation of a serializable struct is target-specific. (By the way, I am not the only one that makes this statement, see section 7 of the attached pdf, it mentions that a CPU target could also benefit from choosing a different layout for a serializable struct).

**Comment by hesingh** [comment]
@hanw Sorry, a typo.  You didn't want to use a header because you wanted to use a construct that can be modified.  Changing any data struct layout is a p4c backend activity internal to your platform.
  
URW is Update Read Write which is a hardware component in the switching asic to modify/insert header(s) of the packet before egress.

**Comment by hesingh** [comment]
I read the serializer.pdf doc.  What is missing in the doc is justification for why explicit serializable struct is inserted in the P4-16 language.  I see the text related to padding and internal bus may need flattened data.  However, it should be noted, these are platform-specific p4c backend issues.  Why should the language change for a platform internals?  

When adding nested structs to P4-16, it is implicit that only emittable nested structs are allowed a header.      
 Why doesn't that solve your issues?  Mihai has already asked "What if we call the new sizeof as DeclaredSizeInBits"?  Perhaps we should take up serializable struct and nested struct in another Issue (file a new one if you have to) because this is a sizeof issue which may be resolved if you replied to Mihai's question.

**Comment by hanw** [comment]
Would you agree that these terms: 'emittable structs', 'bit-vector struct', 'serializable struct', are basically talking about the same thing? It's struct with only 'bit<w>', 'int<w>', 'ser enum', 'bool' fields.

Now the argument is whether we should have a syntax of
`serializable struct { bit<8> f0, bit<9> f1}` to explicitly identify these specialized struct in the IR 
or use 
`struct { bit<8> f0, bit<9> f1}`, and write a few helper functions in the compiler frontend to infer if a struct is also a bit-vector struct / serializable struct. 

I don't see much difference in there. 

Regarding why I am against on substituting sizeof with constant in the frontend, my point was if the sizeof method call is replaced with a constant. How would I write code to adjust the constant value if my tool decides to adjust the layout of the specialized struct. What IR node should I look for to do the Transformation? IR::Constant? How do I know the IR::Constant node is generated from the sizeof method call?

**Comment by hesingh** [comment]
> Regarding why I am against substituting sizeof with constant in the frontend, my point was if the sizeof method call is replaced with a constant. How would I write code to adjust the constant value if my tool decides to adjust the layout of the specialized struct. What IR node should I look for to do the Transformation? IR::Constant? How do I know the IR::Constant node is generated from the sizeof method call?

Please note the use of new type.sizeBits() is an AssignmentStatement.  For example,

bit<32> len = hdr.sizeBits();

In the above statement, len is a local variable, and the MidEndLast  IR which a backend sees, does not even include any local variables.  

The other use is how theojepsen (who filed this issue) has shown.

```
action set_len() {
    hdr.ipv4.totalLen = len(hdr.ipv4) + len(hdr.udp) + len(hdr.my_proto);
}
```

The right side of the AssignmentStatement is changed to a IR:Constant.  The left side saves the constant in the hdr.ipv4 data structure.  Again, your backend should not have any impact with such an AssignmentStatement  statement.  Thus, Mihai is correct:  "If in your architecture headers change sizes and the user needs to know when and where you should just define an architecture-specific function for the user to find out."

**Comment by hesingh** [comment]
> Would you agree that these terms: 'emittable structs', 'bit-vector struct', 'serializable struct', are basically talking about the same thing? It's struct with only 'bit', 'int', 'ser enum', 'bool' fields.
> 

Yes.  

> Now the argument is whether we should have a syntax of
> `serializable struct { bit<8> f0, bit<9> f1}` to explicitly identify these specialized struct in the IR
> or use
> `struct { bit<8> f0, bit<9> f1}`, and write a few helper functions in the compiler frontend to infer if a struct is also a bit-vector struct / serializable struct.
> 
> I don't see much difference in there.

With the nested structs frontend changes, it is implicit that a header can never include a nested struct which is not emittable.  A header is the only entity supported by extract().  Why not use this emittable header for your internal use? As I said before, at packet ingress into the asic, copy the packet header and use the copy.   Additionally, a struct can include several headers where each header is emittable.  Do what  you want internally with this struct which is identifiable because it only has headers.  The frontend already includes a helper function to check if a struct is a bit-vector-only struct and hence serializable.  If you need any other specific checks, sure, the frontend can add another helper.

**Comment by hanw** [comment]
> Please note the use of new type.sizeBits() is an AssignmentStatement. For example,
> 
> bit<32> len = hdr.sizeBits();
> 
> In the above statement, len is a local variable, and the MidEndLast IR which a backend sees, does not even include any local variables.

The sizeBits() IR::MethodCallExpression can be used in any IR::Expression. For instance, it could be used to update the value of user_meta, which is not local variable.

```
user_meta.pkt_len = intrinsic_md.pkt_len - hdr.sizeBits();

// where, intrinsic_md.pkt_len counts total packet length.
// hdr.sizeBits() counts serializable_struct length;
// user_meta.pkt_len counts packet length without serializable_struct.
```

> Your backend should not have any impact with such an AssignmentStatement statement. Thus, Mihai is correct: "If in your architecture headers change sizes and the user needs to know when and where you should just define an architecture-specific function for the user to find out."

In my case, the specialized struct can change sizes, BUT, the user does not need to know about the change.

**Comment by hesingh** [comment]
The example you gave is already covered by the 2nd use case I gave in my note.   Any AssignmentStatement has any IR::Expression and the right side is not a local variable.  Such a statement will make it to the backend in MidEndLast IR. 

I don't fully understand your last statement.  I understand the specialized struct can change sizes.  Are you good with the language adding sizeof and implementing sizeof in the frontend?

**Comment by hanw** [comment]
> Such a statement will make it to the backend in MidEndLast IR.

@hesingh I am not sure which statement you are referring to? 
`hdr.ipv4.totalLen = len(hdr.ipv4) + len(hdr.udp) + len(hdr.my_proto);` ?

> I don't fully understand your last statement. I understand the specialized struct can change sizes. Are you good with the language adding sizeof and implementing sizeof in the frontend?

I am afraid not, but it's not your fault. We just don't have a clear definition on what the size of a struct is. Currently the pass to eliminate nested struct assumes if a struct is used in a header, the size of struct is the sum of all fields sizes in the struct. (Because the flattenHeaders pass removes the nested struct in a header, and the size of a header is computed by summing the size of all its fields). In other case, say if a struct used standalone, the size of struct is undefined.

This seems inconsistent to me, and my gut feeling is something is not quite right.

**Comment by jafingerhut** [comment]
It might seem a bit odd, but if we think these statements make sense:

+ a header must always have a well defined and portable wire format for emit and extract purposes, and thus a well defined and portable definition for sizeBits and sizeBytes (thus size operations on headers can always be implemented and replaced with compile time constants early in the compiler's passes), even if it contains structs

+ a struct need not have a well defined and portable wire format for emit and extract purposes, and thus need not have a well defined and portable definition for sizeBits and sizeBytes (thus size operations on structs should always be implemented in compiler back end), whether or not it contains structs or headers.  (one exception that could be grandfathered in, because the P4_16 already defines it -- emit is allowed on structs that contain _nothing except_ header, header stack, and header_union members).

That might seem inconsistent with structs inside of headers, but if this is called out explicitly to P4_16 programmers, it seems to make a kind of sense to me -- it is whether the _top level_ thing you are doing emit/extract/sizeBit/sizeBytes on that determines whether it is portable across targets (header) vs. target-specific (struct)

**Comment by hesingh** [comment]
> > Such a statement will make it to the backend in MidEndLast IR.
> 
> @hesingh I am not sure which statement you are referring to?
> `hdr.ipv4.totalLen = len(hdr.ipv4) + len(hdr.udp) + len(hdr.my_proto);` ?
> 
Yes, the above one it is and it's the 2nd use case.  Note the v1model.p4 passes in the parsed hdr to controls and the deparser.  This is why the hdr is not a local variable. 

> > I don't fully understand your last statement. I understand the specialized struct can change sizes. Are you good with the language adding sizeof and implementing sizeof in the frontend?
> 
> I am afraid not, but it's not your fault. We just don't have a clear definition on what the size of a struct is. Currently the pass to eliminate nested struct assumes if a struct is used in a header, the size of struct is the sum of all fields sizes in the struct. (Because the flattenHeaders pass removes the nested struct in a header, and the size of a header is computed by summing the size of all its fields). In other case, say if a struct used standalone, the size of struct is undefined.

Would it help if your backend's p4c build has the midend.cpp file will no inclusion of FlattenHeaders pass?  As a result, your backend receives a nested struct in a header - can your backend handle such a header.  If so, it will work. 
> 
> This seems inconsistent to me, and my gut feeling is something is not quite right.

**Comment by hanw** [comment]
@jafingerhut In response to your comment from a couple days ago.
> A struct like you describe _could_ be emitted and extracted, if we define a way for it to happen.

and the comment today.

> a struct need not have a well defined and portable wire format for emit and extract purposes, and thus need not have a well defined and portable definition for sizeBits and sizeBytes (thus size operations on structs should always be implemented in compiler back end)

> It is whether the _top level_ thing you are doing emit/extract/sizeBit/sizeBytes on that determines whether it is portable across targets (header) vs. target-specific (struct)

If we were to redefine emit() and extract() to be allowed on struct, we will need to figure out how to deal with validity bit. By definition, the emit() and extract() methods checks and modifies the validity bits. There is no validity bit on struct. Currently the only way to emit a struct is to wrap it inside another header type which provides the validity bit. However, as soon as we put the struct inside the header, the size of the struct is defined and no longer 'target-specific'.

**Comment by hesingh** [comment]
I think, the theme should be to compute sizeBits/sizeBytes in the p4c frontend, especially for an emittable nested struct inside header.  For whatever reason, if another computation is needed in the backend, do it.

Emit allows a struct and emits the struct recursively since there is no Valid bit to check.   

Changing extract to support a struct is an interesting to-be-specified activity.

**Comment by hesingh** [comment]
Here is a crack at to-be-specified behavior for extract because a header has emittable nested struct.

Extract doesn't support extracting specific bit(s) inside a header.  Hence, if one allows extracting a struct inside a header, how do the other bit-fields inside the header get extracted?  Why bother changing extract to extract bits as well.  Thus, we should not change extract to support a struct.  Just extract the whole header and then it's also legal to set the isValid bit.   In fact, this is what p4c does today, even with nested structs code checked in.  

If anyone has a critical need to extract a struct inside a header, please speak up.

Andy and other folks made the sound choice to support sizeof, only for a header.  However, why not extend sizeof behavior to also support sizeof for an emittable struct?  Sizeof just adds a check "if arg is struct, check if struct is emittable, return sizeof else return failure".  This way's Han's request is addressed.  

I agree with Andy's note today for rest of the things he has said.

**Comment by hanw** [comment]
I wouldn't draw conclusions too quickly on this issue. I feel we need a type system expert to chime in and/or work out the type system of P4-16 before we can make more decisions. Otherwise, we may start making wrong decisions that are hard to revert in the long term and hurt the language.

**Comment by hesingh** [comment]
Sure, sounds good to wait.  

What I have said in my note is a summary of what the p4c code supports today and why, including what a rat's nest exists changing the arg type of extract.  Then, my note, adds a change to allow sizeof for an emittable struct.

**Comment by jafingerhut** [comment]
@hanw  Sure, one way to define emit on struct is "always emit unconditionally, as if it were a valid header".  And a way to define extract on struct is "just like on a header, except there is no valid bit to set".  If one wanted to tell the difference between an extract succeeding on a struct, vs. failing, they could create their own custom valid bit in user-defined metadata and assign that value a 1 after the extract call on the struct.

I am not saying I know there are 0 wrinkles or gotchas remaining with making all structs support extract/emit/sizeBits/sizeBytes, but it certainly seems possible to do it, and becomes _much_ easier to get agreement on across multiple implementers if it is defined to be target-specific in format and size.

**Comment by jafingerhut** [comment]
To be very explicit, it seems to me like this program should be pretty easy to explain to someone what is going on, and could/should be legal P4, if we define emit/extract on structs to be target-specific:

```
struct s1_t {
    bit<4> a;
    bit<2> b;
}

header h1_t {
    s1_t x;
    s1_t y;
    bit<4> z;
}

struct headers_t {
    h1_t h1;
}

// ... later ...

    // These are defined in scope of deparser code:

    // headers_t hdr
    // s1_t s1;
    // s2_t s2;

    pkt.emit(hdr.h1);  // emit h1 if h1.isValid()is true.  It is 16 bits long, target-independent format
        // Why are the two fields of type s1_t target-independent format?
        // Because they are inside of a header definition.
        // No other reason, but that is enough of a reason.

    pkt.emit(s1);     // emit s1 unconditionally.  Its bit level format and size is target-dependent.

    pkt.emit(hdr.h1.y);   // emit hdr.h1.y unconditionally.  Its format and size are target-dependent.
        // Why is target-dependent format allowed?
        // Because the emit is on a value of type struct, not a header,
        // even though that struct happens to be part of a header.
        // A target is free to emit it in the same bit order as when it is
        // inside of header, if the target implementers choose to do so, but they need not.

    pkt.emit(s2);    // emit s2 unconditionally.  Its bit level format and size are target-dependent
```

**Comment by hesingh** [comment]
@hanw Have you tested the sizeBits/sizeBytes PR changes with the Tofino p4c backend?

https://github.com/p4lang/p4c/pull/1720

The single file (typeChecker.cpp) can be copied from this repo:  https://github.com/hesingh/p4c

I am trying to understand if  any p4c backend has issues with the PR.   thanks.

**Comment by hanw** [comment]
@jafingerhut Please look at the following two cases. 

Case 1:
```
struct s1_t {
    bit<4> a;
    bit<2> b;
}

struct s2_t {
    s1_t x;
}  

header h1_t {
    s1_t x;
    s1_t y;
}

// These are defined in scope of deparser code:
// headers_t hdr;


// assuming emit on struct is target-specific, emit on header is target-independent.
// is it correct to say parser received bit<4> x.a, bit<2> x.b, bit<4> y.x.a, bit<2> y.x.b packed together.
pkt.emit(hdr);  

// and parser receives a target-specific packing with the following two emits?
// e.g., could be bit<4> x.a, bit<2> x.b, bit<2> padding_0, bit<4> y.x.a, bit<2> y.x.b, bit<2> padding_1.
pkt.emit(hdr.x);
pkt.emit(hdr.y);
```

Case 2:
```
struct s1_t {
    bit<4> a;
    bit<2> b;
}

struct s2_t {
    s1_t x;
}  

header h1_t {
    bit<8> type;
    s1_t x;
    s1_t y;
}

// These are defined in scope of deparser code:
// headers_t hdr;

// only the following emit is supported.
pkt.emit(hdr);  

// the following code is illegal.
pkt.emit(hdr.type);  // disallowed, because bit<n> cannot be emitted.
pkt.emit(hdr.x);
pkt.emit(hdr.y);

// there is no way to emit x, y in a target-dependent way in case 2.
// unfortunately, be able to emit struct in header in a target-dependent way is a common use case for us.
```

**Comment by hanw** [comment]
My question is: how to emit struct x and y with a target-dependent layout in case 2?

**Comment by jafingerhut** [comment]
What _stops_ you from writing a P4 compiler that emits struct x and y with a target-dependent layout in case 2?

**Comment by hanw** [comment]
Right now, nothing. however, if the specification goes: a header is always Packed, even if it contains structs. A struct is never Packed, even if it happens to be included as part of a header. Then the spec would stop me.

**Comment by mbudiu-vmw** [comment]
If you want custom serialization just define your own packet_out-like extern and use that to emit.

**Comment by hanw** [comment]
I rather not doing that, it feels wrong to invent new extern, when we already have emit/extract in the language.

**Comment by jafingerhut** [comment]
How would the spec stop you?

The emit on the header could be flattened to a single flat header if that helps you think about the compiler back end implementation.  The struct is no longer there in the back end/implementation.

The emit on the struct would be handled independently and separately by the compiler.  It is free to take those fields and pad/reorder/restructure them however the compiler back end chooses.  It need not affect the emit on the header at all.

**Comment by jafingerhut** [comment]
In other words, if you are thinking that the definition and implementation of packet.emit(hdr.h1) where h1 is a header type _must_ be implemented and defined as "loop through the members, and call emit on each of them in order", then please stop thinking that it must be this way.  packet.emit(hdr.h1) is one user-invoked statement, and how that is implemented internally can be completely independent of how packet.emit(hdr.h1.s1) is implemented where hdr.h1.s1 is a struct type.

**Comment by mbudiu-vmw** [comment]
That's what externs are for: Defining target-specific behaviors.

**Comment by hanw** [comment]
> That's what externs are for: Defining target-specific behaviors.

I think there is a chance we can resolve the issue without creating extern as a escape. That's why I am pushing for it.

**Comment by jafingerhut** [comment]
OK, I think we may be getting our minds more closely aligned, then.  Now to convince the LDWG :-)

Personally, I know it would be a backwards-incompatible change if we did this, but if we _removed_ the special case in the P4_16 spec today where emit on a struct that contained _only_ header, header stack, and header_union members, so that we could change the spec to say 'emit on a struct is _always_ target-dependent in its wire format, even in that case' it would make for fewer cases to worry about.  But it may not be considered a big complexity to keep that special case around, too.

**Comment by hanw** [comment]
100% agreed with you!

**Comment by jafingerhut** [comment]
I would even propose a slight change to what you said above.  You said: "For emitting struct with nested struct and no headers, the behavior could be target-dependent."

I would propose:

For emitting a struct that contains _only_ header, header stack, and header_union members, emit has target-independent behavior as specified already in the language spec.

For emitting _any_ other kind of struct, e.g. it contains 3 headers and one bit<7>, now emit becomes target-dependent, even though it contains some header members.

So every struct is either the current supported kind, or "any other struct is target-dependent".

**Comment by hanw** [comment]
Another catch is we need to define emit operation on `bit<n>`, `int<n>` and `varbit<n>`. Otherwise, the definition for emit does not make sense.

**Comment by hanw** [comment]
We don't need to support the syntax `pkt.emit(f0)` where `f0` is `bit<8>`, but we need to say something about what happens with emitting bits.

**Comment by jafingerhut** [comment]
I mean, we _could_ define what an emit does on those types, but it isn't _necessary_.  We can say that emit is defined on structs, and "the implementation does whatever it needs to to preserve the contents of the struct in a target-dependent way", yes?

In other words, why do you believe we _need_ to say something about what happens when emitting bits?

**Comment by jafingerhut** [comment]
Let me add a detail that might be important here.  I am assuming the following.

Suppose we emit a struct that in a target-dependent format, e.g. this struct:
```
struct s1_t {
    bit<4> f1;
    bit<7> f2;
    bit<3> f3;
}
```

Later we extract a variable with _that same struct type s1_t_.  The extract should always result in the extracted variable being equal (via P4_16's `==` operator) as the emitted struct variable.

However, suppose we later instead do not do an extract on a struct with type s1_t, but do _two_ extract operations, first on struct s2_t, then on s3_t, defined like this:
```
struct s2_t {
    bit<4> f1;
    bit<7> f2;
}
struct s3_t {
    bit<3> f3;
}
```
Is any behavior guaranteed by doing these 2 extracts, after that emit?  I would so _NO_.  The target-dependent format for s2_t, followed by the target-dependent format for s3_t, _NEED NOT_ be the same as the target-dependent format for s1_t, even for the same target.  A programmer should not rely on that.

Why not?  The emit could have emitted the fields in the reverse order from the order the fields were defined in s1_t, or the same order, or some arbitrary order.  It might have put some of the bits for field f3 between the bits for f1 and f2, and the rest at the end.  You just don't know, and should never rely upon it.

The only reason you should ever emit a struct, is if you plan later to extract that precise same struct type, on the same target device, at the same "byte offset" from the beginning of the packet as the emit occurred.  Anything else, and you have just pointed the gun at your foot and pulled the trigger very, very confidently and foolishly.

**Comment by jafingerhut** [comment]
Both p4c and the P4_16 language spec have been updated to add two methods on headers: minSizeInBytes() and minSizeInBits().  This proposed syntax in the original comment:
```
action set_len() {
    hdr.ipv4.totalLen = len(hdr.ipv4) + len(hdr.udp) + len(hdr.my_proto);
}
```
can thus now be implemented using this syntax:
```
action set_len() {
    hdr.ipv4.totalLen = hdr.ipv4.minSizeInBytes() + hdr.udp.minSizeInBytes() + hdr.my_proto.minSizeInBytes();
}
```
